awscli/customizations/wizard/core.py

```diff
@@ -0,0 +1,237 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Core planner and executor for wizards."""
+import jmespath
+from botocore import xform_name
+
+
+class Planner(object):
+    _DONE_STEP = 'DONE'
+    _STOP_RUNNING = object()
+
+    def __init__(self, step_handlers):
+        self._step_handlers = step_handlers
+        # These will hold the values we store from
+        # running through the plan steps.
+        self._parameters = {}
+
+    def run(self, plan):
+        self._parameters.clear()
+        steps_iter = self._steps(plan)
+        step = next(steps_iter)
+        while step is not self._STOP_RUNNING:
+            next_step_name = self._run_step(step)
+            step = steps_iter.send(next_step_name)
+        return self._parameters
+
+    def _steps(self, plan):
```

Comment:
Not a strong reason.  The main reason is that all the state variables related to keep tracking of the step transitions are local to the `_steps` method.  Otherwise you'd either have to mix them in as instance vars on the planner class or create a separate iterator object that tracks the state.